File: clone.py
import os
import subprocess

def main():
    github_user = input("Enter your GitHub username: ").strip()
    clone_dir = input("Enter the directory where you want to clone the repos: ").strip()

    os.makedirs(clone_dir, exist_ok=True)
    os.chdir(clone_dir)

    try:
        result = subprocess.run(
            [
                "gh", "repo", "list", github_user, "--limit", "100", "--json",
                "nameWithOwner,isFork,isArchived", "--jq",
                ".[] | select(.isFork == false and .isArchived == false) | .nameWithOwner"
            ],
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True
        )

        if result.returncode != 0:
            print("Failed to fetch repositories. Make sure 'gh' is installed and configured.")
            print(result.stderr)
            return

        repos = result.stdout.strip().splitlines()
    except Exception as e:
        print(f"Error fetching repositories: {e}")
        return

    for repo in repos:
        repo = repo.strip()

        repo_name = repo.split("/")[1] if "/" in repo else repo
        repo_dir = repo_name.replace("/", "_")  # Replace '/' with '_' to avoid invalid directory names
        clone_path = os.path.join(clone_dir, repo_dir)

        os.makedirs(clone_path, exist_ok=True)

        clone_command = f"git clone git@github.com:{repo} {clone_path}"
        try:
            result = subprocess.run(
                ["sh", "-c", clone_command],
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True
            )

            if result.returncode == 0:
                print(f"{repo} has been cloned into {clone_path}")
            else:
                print(f"Failed to clone {repo} into {clone_path}")
                print(result.stderr)
        except Exception as e:
            print(f"Error cloning {repo}: {e}")

    print(f"All repos have been cloned into {clone_dir}")

if __name__ == "__main__":
    main()

File: clone.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"sync"
)

func main() {
	user := flag.String("user", "", "GitHub username")
	dir := flag.String("dir", "", "Directory to clone repos into")
	workers := flag.Int("workers", 3, "Number of concurrent clone operations")
	flatClone := flag.Bool("flat", false, "Clone repos to parent folder (no owner subdirectory)")
	flag.Parse()

	githubUser := *user
	cloneDir := *dir

	if githubUser == "" {
		githubUser = promptUser("Enter your GitHub username: ")
	}
	if cloneDir == "" {
		cloneDir = promptUser("Enter the directory where you want to clone the repos: ")
	}

	cloneToParent := *flatClone
	if !cloneToParent {
		cloneToParent = promptYesNo(
			"Clone to parent folder without owner subdirectory? (y/n): ",
		)
	}

	if githubUser == "" || cloneDir == "" {
		fmt.Println("Error: GitHub username and clone directory are required")
		os.Exit(1)
	}

	if err := os.MkdirAll(cloneDir, 0755); err != nil {
		fmt.Printf("Failed to create directory: %v\n", err)
		os.Exit(1)
	}

	repos, err := fetchRepositories(githubUser)
	if err != nil {
		fmt.Printf("Failed to fetch repositories: %v\n", err)
		os.Exit(1)
	}

	if len(repos) == 0 {
		fmt.Println("No repositories found")
		return
	}

	fmt.Printf("Found %d repositories. Starting clones with %d workers...\n",
		len(repos), *workers)

	semaphore := make(chan struct{}, *workers)
	var wg sync.WaitGroup

	for _, repo := range repos {
		wg.Add(1)
		go func(r string) {
			defer wg.Done()
			semaphore <- struct{}{}
			defer func() { <-semaphore }()

			cloneRepository(r, cloneDir, cloneToParent)
		}(repo)
	}

	wg.Wait()
	fmt.Printf("Completed! All repos have been cloned into %s\n", cloneDir)
}

func promptUser(prompt string) string {
	reader := bufio.NewReader(os.Stdin)
	fmt.Print(prompt)
	input, _ := reader.ReadString('\n')
	return strings.TrimSpace(input)
}

func promptYesNo(prompt string) bool {
	reader := bufio.NewReader(os.Stdin)
	fmt.Print(prompt)
	input, _ := reader.ReadString('\n')
	answer := strings.TrimSpace(strings.ToLower(input))
	return answer == "y" || answer == "yes"
}

func fetchRepositories(githubUser string) ([]string, error) {
	cmd := exec.Command(
		"gh", "repo", "list", githubUser,
		"--limit", "100",
		"--json", "nameWithOwner,isFork,isArchived",
		"--jq", ".[] | select(.isFork == false and .isArchived == false) | .nameWithOwner",
	)

	output, err := cmd.CombinedOutput()
	if err != nil {
		return nil, fmt.Errorf("gh command failed: %w", err)
	}

	var repos []string
	for _, line := range strings.Split(strings.TrimSpace(string(output)), "\n") {
		if trimmed := strings.TrimSpace(line); trimmed != "" {
			repos = append(repos, trimmed)
		}
	}
	return repos, nil
}

func cloneRepository(repo, baseDir string, flatClone bool) {
	var clonePath string

	if flatClone {
		repoName := filepath.Base(repo)
		clonePath = filepath.Join(baseDir, repoName)
	} else {
		clonePath = filepath.Join(baseDir, repo)
	}

	if err := os.MkdirAll(clonePath, 0755); err != nil {
		fmt.Printf("❌ Failed to create directory for %s: %v\n", repo, err)
		return
	}

	cmd := exec.Command("git", "clone", fmt.Sprintf("git@github.com:%s", repo),
		clonePath)

	if output, err := cmd.CombinedOutput(); err != nil {
		fmt.Printf("❌ Failed to clone %s: %v\n", repo, err)
		if len(output) > 0 {
			fmt.Printf("   %s\n", string(output))
		}
	} else {
		fmt.Printf("✅ %s cloned successfully\n", repo)
	}
}